feat: 云·绝区零

antecanis8 opened the pull request:
#983 
<img width="1369" height="924" alt="image" src="https://github.com/user-attachments/assets/c44ceec3-287b-4a74-939a-e6d5dae4e6b4" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## 发布说明

* **新功能**
  * 新增云游戏客户端支持，用户可在本地游戏和云游戏间灵活切换
  * 新增云游戏排队系统，支持队列人数和预计等待时间识别
  * 新增云游戏设置选项，包括邦邦点快速队列偏好选择
  * 优化鼠标交互响应能力，提升云游戏环境下的操作体验

* **文档**
  * 新增云游戏适配开发指南
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

antecanis8 committed:
feat: 添加了对云绝区零配置和路径的支持

antecanis8 committed:
feat: 添加国服PC云游戏的排队和进入游戏功能

kawayiYokami reviewed (commented):
你好，这是一个来自红豆沙的哈基米的审查结果：

整体来看，这是一个非常棒的PR，功能实现得很完整，代码结构也很清晰。但在细节上发现了一些可以改进的地方：

### 潜在问题

1.  **窗口标题硬编码**：在 src/zzz_od/context/zzz_context.py 中，云游戏的窗口标题被硬编码为 '云·绝区零'，并且代码中的 TODO 也提到了，这会导致在国际服上无法正确识别窗口。
2.  **客户端类型判断错误**：在 src/zzz_od/operation/enter_game/enter_game.py 中，用于判断客户端类型的代码是 self.ctx.game_account_config.game_client == 'cloud'。根据 game_account_config.py 中的定义，正确的枚举值应该是 GameClientTypeEnum.CLOUD_GAME.value.value (即 'CLOUD_GAME')，直接使用 'cloud' 字符串可能导致此判断永远不为真。

### 改进建议

1.  **代码复用**：在 src/zzz_od/operation/enter_game/enter_game.py 的 cn_pc_cloud_queue 函数中，识别
排队人数和预计等待时间的两块OCR逻辑几乎完全相同。可以考虑将这部分提取为一个独立的辅助函数，以减少重复代码。

希望这些建议能有所帮助！

antecanis8 commented on src/one_dragon/base/config/game_account_config.py:
```diff
@@ -31,6 +36,8 @@ def __init__(self, instance_idx: int,
                  default_platform: Optional[str] = None,
                  default_game_region: Optional[str] = None,
                  default_game_path: Optional[str] = None,
+                 default_cloud_game_path: Optional[str] = None,
```
因为我希望用户可以在本地游戏和云游戏之间切换，而不是每次切换之后都要手动选择路径

Qian23333 replied:
@ setter
路径
if 本地
set 本地
else

A-nony-mous commented on src/zzz_od/operation/back_to_normal_world.py:
```diff
@@ -103,6 +103,13 @@ def check_screen_and_run(self) -> OperationRoundResult:
             op.execute()
             return self.round_retry(result, wait=1)
 
+        #云绝区零 排队
```
打游戏是打游戏，上号是上号

Qian23333 replied:
可以删了

antecanis8 replied:
done

antecanis8 committed:
fix: 云游戏排队功能移动到EnterGame之前并优化排队流程

新增 CloudGameQueue 操作类，用于处理国服PC云游戏的排队逻辑，包括OCR识别排队人数和预计等待时间，并在适当时机进入游戏。 撤销了对 EnterGame 类的大部分更改。同时调整了OpenAndEnterGame 类，使其能够正确识别和处理云游戏场景，避免重复登录操作。

- 在 `one_dragon/base/operation/operation.py` 中新增 `round_by_ocr_text` 方法，支持OCR文本识别并返回识别结果功能
- 新增 `zzz_od/operation/enter_game/cloud_game_queue.py` 文件，实现完整的云游戏排队逻辑
- 修改 `zzz_od/operation/enter_game/enter_game.py`，移除原有云游戏相关逻辑，改为通过云游戏客户端类型判断是否跳过登录
- 更新 `zzz_od/operation/enter_game/open_and_enter_game.py`，在进入游戏前增加云游戏排队处理逻辑
- 注释掉 `GamePlatformEnum` 中的模拟器配置项（暂不使用）

antecanis8 committed:
fix: 修复云游戏排队模块导致的循环导入

antecanis8 committed:
Merge branch 'main' into cloud

antecanis8 committed:
fix: 修复云游戏排队模块中的缩进错误和OCR缓存调用函数错误

修复了 `cloud_game_queue.py` 中的代码缩进问题，以及 `operation.py` 中 OCR 结果获取方法名错误的问题。

Qian23333 commented on assets/game_data/screen_info/enter_game.yml:
py文件已经独立了，分开吧

Qian23333 commented on src/one_dragon_qt/view/setting/setting_instance_interface.py:
```diff
@@ -180,10 +182,18 @@ def _get_instanceSwitch_group(self) -> QWidget:
 
-        self.game_path_opt = PushSettingCard(icon=FluentIcon.FOLDER, title='游戏路径', text='选择')
+        self.game_client_type_opt = ComboBoxSettingCard(icon=FluentIcon.GAME, title='游戏客户端',
+                                                        options_enum=GameClientTypeEnum)
```
按照上面的评论修改，组件可以复用

Qian23333 commented on src/zzz_od/context/zzz_context.py:
```diff
@@ -166,13 +166,15 @@ def init_by_config(self) -> None:
                 from one_dragon.base.config.game_account_config import GameRegionEnum
                 win_title = '绝区零' if self.game_account_config.game_region == GameRegionEnum.CN.value.value else 'ZenlessZoneZero'
+                if self.game_account_config.game_client == GameClientTypeEnum.CLOUD_GAME.value.value:
+                    win_title = '云·绝区零' #TODO 外国的云游戏叫什么？
```
没出

antecanis8 replied:
done

Qian23333 commented on src/zzz_od/operation/enter_game/cloud_game_queue.py:
```diff
@@ -0,0 +1,112 @@
+        # 仅作为中转，直接切换回第一个节点继续处理
+        return self.round_wait(status='排队中转', wait=1)
+
+def __debug():
+    ctx = ZContext()
+    ctx.init_by_config()
+    ctx.start_running()
+    ctx.init_ocr()
+    op = CloudGameQueue(ctx)
+    op.execute()
+
+
+if __name__ == '__main__':
+    __debug()
```
nitpick

Qian23333 commented on src/zzz_od/operation/enter_game/open_and_enter_game.py:
```diff
@@ -42,5 +42,13 @@ def wait_game(self) -> OperationRoundResult:
     def enter_game(self) -> OperationRoundResult:
         from zzz_od.operation.enter_game.enter_game import EnterGame
+        # 如果是云游戏 那么先运行CloudGameQueue 再运行EnterGame 如果不是云游戏 直接EnterGame
+        if self.ctx.game_account_config.game_client == GameClientTypeEnum.CLOUD_GAME.value.value:
```
config里面加个property，返回是否为云游戏

antecanis8 committed:
fix: 补全缺少的截图区域

领取云绝区零每天赠送的15分钟时长 '国服PC云-点击空白区域关闭' 区域

antecanis8 committed:
refactor: 云游戏相关截图区域移动到单独的yml文件

- 新增云游戏屏幕信息配置文件 (cloud_game.yml)
- 调整返回大世界的逻辑以支持云游戏排队
- 移除不必要的TODO注释

antecanis8 committed:
feat: 使用 taskkill 命令关闭游戏进程

通过窗口标题过滤并强制关闭游戏进程，支持关闭云游戏。

antecanis8 commented on assets/game_data/screen_info/enter_game.yml:
```diff
@@ -449,4 +449,4 @@ area_list:
   template_match_threshold: 0.7
   color_range: null
-  goto_list: []
+  goto_list: []
```
我不到啊

A-nony-mous replied:
> 我不到啊

不知道就回撤 git diff 一下 删空格、换行这种不要出现在commit里

这个情况可以不revert 再交一个commit恢复到原来样子就可以（去历史记录里复制原来的地方覆盖过来）

antecanis8 replied:
done

antecanis8 commented on src/one_dragon/base/operation/operation.py:
```diff
@@ -1089,6 +1089,43 @@ def round_by_ocr(
+
+            ocr_result_map = self.ctx.ocr.run_ocr(to_ocr_part)
+
+        # 将所有识别到的文本连接成一个字符串
```
因为我需要的是这一小片区域的识别结果 不会有多个值

Qian23333 replied:
考虑一下先把这个更改提出来先合并，别的地方有要用到的

antecanis8 committed:
feat: 支持使用邦邦点插队进入云游戏

revert: 撤回BackToNormalWorld中对云游戏的识别

antecanis8 committed:
revert: 恢复enter_game.yml末尾的空行

antecanis8 committed:
feat: 添加is_cloud_game属性以判断当前游戏客户端是否为云游戏

antecanis8 committed:
refactor: 修改OCR识别结果返回格式为字典

将operation.py中的OCR识别结果从拼接字符串改为返回原始字典格式，便于后续处理。同时更新cloud_game_queue.py中对OCR结果的处理逻辑，确保兼容新的返回格式。

修复cloud_game_queue.py中等待时间文本识别后变量赋值错误的问题。

antecanis8 committed:
fix: 修复用户没有邦邦点时卡在选择是否使用邦邦点的问题

antecanis8 committed:
fix: 修复云游戏排队时OCR识别预计等待时间错误的问题

antecanis8 committed:
fix: 修改是否在云游戏排队中的OCR区域

antecanis8 committed:
fix: 修复 插队或排队 节点和 排队 节点的路径问题

antecanis8 commented on src/zzz_od/operation/enter_game/cloud_game_queue.py:
```diff
@@ -0,0 +1,139 @@
+    @node_from(from_name='画面识别', status='国服PC云-普通队列')
+    @node_from(from_name='画面识别', status='国服PC云-退出排队')
+    @node_from(from_name='国服PC云-排队中转', status='排队中转')
+    @operation_node(name='国服PC云-排队')
```
done

antecanis8 committed:
feat: 适配畅玩卡的不排队直接进入游戏

antecanis8 committed:
refactor: 避免与 True/False直接比较, 移除未使用代码

antecanis8 committed:
refactor: 移除未使用的导入GameClientTypeEnum

antecanis8 committed:
style: 移除cloud_game.yml的末尾空格

ShadowLemoon committed:
Merge branch 'main' into pr/antecanis8/1308

ShadowLemoon committed:
perf: 优化客户端配置项

ShadowLemoon committed:
perf: 迁移排队操作和配置

ShadowLemoon committed:
feat: 一条龙运行 恢复排队

ShadowLemoon committed:
fix: 添加云排队运行记录

antecanis8 committed:
feat: 识别到大世界-信息按钮时跳过排队流程

antecanis8 committed:
feat: 未识别到国服PC云-切换窗口时跳过云游戏排队流程

revert: 撤回#b3e68c45

ShadowLemoon committed:
feat: 迁移旧路径

ShadowLemoon committed:
fix: 恢复启动游戏未配置路径提示

ShadowLemoon committed:
style

ShadowLemoon committed:
perf: 调整逻辑

antecanis8 committed:
perf: 移除云游戏排队操作中不必要的等待

ShadowLemoon committed:
perf: 简化ocr识别文本的调用

idk500 reviewed (approved):
冲啊

ShadowLemoon committed:
perf: 有需要的时候再导入subprocess

ShadowLemoon committed:
perf: 避免合并冲突

ShadowLemoon committed:
fix: 修复ocr

ShadowLemoon committed:
fix: 参数顺序

ShadowLemoon committed:
fix: 替换ocr函数

ShadowLemoon committed:
fix: 不再使用ocr缓存

ShadowLemoon committed:
fix: 修复v9o

Qian23333 commented on src/one_dragon/base/controller/pc_controller_base.py:
```diff
@@ -188,7 +188,9 @@ def close_game(self):
-            self.game_win.get_win().close()
+            import subprocess
+            win_title = self.game_win.win_title
+            subprocess.run(["taskkill", "/f", "/fi", f"WINDOWTITLE eq {win_title}"], check=True, capture_output=True)
```
原来的方式是发送一个结束信号，云游戏下会弹个是否退出，无法正常退。这个直接强制结束进程

antecanis8 replied:
<img width="1920" height="1080" alt="QQ20250823-0227122" src="https://github.com/user-attachments/assets/f967efb0-a74a-4032-8822-634f41a7c6eb" />
云游戏弹出的选择框

DoctorReid commented on src/zzz_od/operation/enter_game/enter_game.py:
```diff
@@ -32,6 +32,9 @@ def __init__(self, ctx: ZContext, switch: bool = False):
         self.already_login: bool = False  # 是否已经登录了
+        if self.ctx.game_account_config.is_cloud_game:
+            self.force_login = False
+            self.already_login = True
```
逻辑还是散落在比较多地方，需要一个文档整理一下这个登录游戏的流程，不然后续不好维护。文档放在 `docs/develop/ ` 下

Qian23333 commented on src/zzz_od/application/zzz_one_dragon_app.py:
```diff
@@ -40,6 +41,7 @@ def __init__(self, ctx: ZContext):
     def get_app_list(self) -> List[ZApplication]:
         from zzz_od.application.notify.notify_app import NotifyApp
         return [
+            CloudQueueApp(self.ctx),
```
因为此处是一条龙接管排队的app，但是如果一条龙拉起游戏的话，必须要在进入游戏前排队（或者跳过进入游戏的逻辑直接由这边接管）

antecanis8 committed:
feat: 重构 drag_mouse 函数, 直接调用ctypes

antecanis8 committed:
feat: 更新 drag_mouse 函数，复用获取屏幕尺寸的方法

antecanis8 committed:
Merge branch 'main' into cloud

# Conflicts:
#	src/zzz_od/context/zzz_context.py

antecanis8 committed:
docs: 添加云游戏适配文档及相关功能实现

Qian23333 commented:
能否研究一下pyautogui的逻辑并做猴子补丁之类的处理？尽量避免在python使用系统dll吧

antecanis8 commented:
> 能否研究一下pyautogui的逻辑并做猴子补丁之类的处理？尽量避免在python使用系统dll吧

pyautogui也是调用ctypes.windll.user32, 而且不是应该更加避免使用猴子补丁吗

Qian23333 commented on src/one_dragon/base/controller/pc_controller_base.py:
```diff
@@ -260,16 +266,53 @@ def get_mouse_sensitivity():
+    dx = (end.x - start.x) / steps
+    dy = (end.y - start.y) / steps
+
+    # 逐步移动鼠标
```
为什么要逐步？
这里可预见的是会拖慢普通的客户端拖拽速度

antecanis8 replied:
在duration一定的情况下，为什么逐步会让拖拽速度变慢？

ShadowLemoon committed:
Merge branch 'main' into pr/antecanis8/1308

ShadowLemoon committed:
fix: 适配应用组更改

ShadowLemoon committed:
feat: 云游戏作为窗口未准备好的一种状态

ShadowLemoon commented:
taskkill可能会杀掉本地多用户中其他用户的进程，需要测试

antecanis8 committed:
Merge branch 'main' into cloud

antecanis8 committed:
Merge remote-tracking branch 'upstream/main' into cloud

antecanis8 committed:
Merge remote-tracking branch 'upstream/main' into cloud

ShadowLemoon committed:
refactor: 重写触发云游戏排队逻辑
 - check_game_window中添加钩子函数check_game_initialized，支持子类扩展检查逻辑
 - ZOperation类重写钩子函数以支持已经打开窗口的云游戏接续排队

ShadowLemoon committed:
revert: 不需要在控制器层面适配云游戏

ShadowLemoon committed:
feat: open_and_enter_game操作中添加排队节点

antecanis8 committed:
Merge remote-tracking branch 'origin/main' into cloud2

# Conflicts:
#	src/one_dragon/base/controller/pc_controller_base.py
#	src/zzz_od/controller/zzz_pc_controller.py
#	src/zzz_od/gui/view/one_dragon/zzz_one_dragon_setting_interface.py
#	src/zzz_od/operation/enter_game/enter_game.py
#	src/zzz_od/operation/zzz_operation.py

antecanis8 committed:
Refactor game account configuration and cloud queue operations

- Updated `game_account_config.py` to improve code readability by standardizing string quotes and removing unnecessary lines.
- Added `prefer_bangbang_points` property to `GameAccountConfig` for better configuration management.
- Refactored `setting_instance_interface.py` to enhance UI settings for game accounts, including the addition of a toggle for `prefer_bangbang_points`.
- Removed obsolete cloud queue application files and related configurations to streamline the codebase.
- Introduced `CloudGameQueue` operation to handle cloud gaming queue logic, including screen checks and queue management.
- Enhanced `OpenAndEnterGame` operation to integrate cloud game queue handling.
- Updated `ZOperation` to include checks for cloud game initialization and window readiness, improving overall game launch reliability.

antecanis8 committed:
fix: 切换游戏客户端选择后及时重启控制器

antecanis8 committed:
fix: 修复EnterGame检查游戏窗口导致递归

antecanis8 committed:
feat: 添加通过窗口句柄关闭游戏的功能

antecanis8 committed:
feat: ZApplication覆写 check_game_window()

antecanis8 committed:
docs: 更新云游戏适配文档

antecanis8 committed:
style: 恢复被ruff format修改的代码

ShadowLemoon commented on src/one_dragon_qt/view/setting/setting_instance_interface.py:
```diff
@@ -289,6 +296,19 @@ def _get_instanceSwitch_group(self) -> QWidget:
+        self.prefer_bangbang_points_opt = SwitchSettingCard(
+            icon=FluentIcon.SPEED_HIGH,
+            title='邦邦点快速队列',
+            content='云游戏排队时优先使用邦邦点快速队列',
```
邦邦点是云绝特有的 不进框架

antecanis8 replied:
那放在哪里 之前还有排队app的时候是放在app设置里的
现在这个挨着游戏客户端 云游戏加的两个设置是放在一起的
<img width="1425" height="950" alt="image" src="https://github.com/user-attachments/assets/937b5287-29d4-4b39-992c-024b60a511dd" />

ShadowLemoon replied:
游戏设置吧

antecanis8 replied:
done

ShadowLemoon commented on src/zzz_od/operation/zzz_operation.py:
```diff
@@ -25,3 +35,32 @@ def __init__(self, ctx: ZContext,
                            need_check_game_win=need_check_game_win,
                            op_to_enter_game=op_to_enter_game)
+
+    def check_game_initialized(self) -> OperationRoundResult:
```
我留这个是作为钩子函数的 不需要再重写下面这个check_game_window方法重复判断窗口就绪 直接重写这个加云绝判定

antecanis8 replied:
done

ShadowLemoon commented on src/zzz_od/operation/zzz_operation.py:
```diff
@@ -2,11 +2,21 @@
 
 
 class ZOperation(Operation):
+    CLOUD_GAME_NOT_ENTERED_AREA_LIST: list[tuple[str, str]] = [
```
抽个ZZZCloudMixin

ShadowLemoon replied:
self.ctx: ZContext = ctx
op_to_enter_game = OpenAndEnterGame(ctx)
这两个赋值也丢进mixin里，叫ZOperationMixin

ShadowLemoon replied:
写property可能比较合适，不需要给mixin做init

antecanis8 replied:
done

ShadowLemoon commented on src/zzz_od/application/zzz_application.py:
```diff
@@ -38,3 +48,24 @@ def __init__(self, ctx: ZContext, app_id: str,
         self.ctx.controller.active_window()
         Application.handle_resume(self)
+
+    def check_game_window(self) -> OperationRoundResult:
```
同ZOPERATION

ShadowLemoon replied:
这个函数整个都可以不要

antecanis8 replied:
done

ShadowLemoon replied:
你也没删啊

ShadowLemoon commented on docs/develop/cloud_game_adaptation.md:
```diff
@@ -0,0 +1,154 @@
+
+云游戏排队不作为独立 `Application` 注册，排队能力内置在 `OpenAndEnterGame` 的 `云游戏排队` 节点中。普通本地游戏不会执行该节点的排队逻辑。
+
+### 主要逻辑 (`cloud_game_queue.py`)
```
这里没更新

antecanis8 committed:
Merge remote-tracking branch 'origin/main' into cloud2

antecanis8 committed:
refactor: 重构ZApplication和ZOperation类，移除冗余代码并引入Mixin

ShadowLemoon commented on src/zzz_od/operation/zzz_operation_mixin.py:
```diff
@@ -0,0 +1,56 @@
+        self._op_to_enter_game = value
+
+
+class ZZZCloudMixin:
```
这个可以合到上面去

antecanis8 replied:
done

ShadowLemoon commented on src/zzz_od/controller/zzz_pc_controller.py:
```diff
@@ -15,6 +18,7 @@ class ZPcController(PcControllerBase):
     def __init__(
             self,
             game_config: GameConfig,
+            win_title: str,
```
我印象里我是把标题从构造函数里去掉了的

antecanis8 replied:
done

antecanis8 committed:
refactor: 移除ZZZCloudMixin类, 合并到ZOperationMixin

- 还原对ZPcController的修改

antecanis8 committed:
refactor: 云游戏的邦邦点相关设置迁移到游戏设置

antecanis8 committed:
fix: 修复新的拖动操作在多屏幕设备偏移的问题

当游戏窗口在非主屏，尤其是副屏在主屏左侧或上方导致桌面坐标为负数时，绝对坐标会被按主屏归一化，拖拽目标可能偏移、裁剪或落到错误屏幕。

修复方向是按虚拟桌面矩形归一化，并带上 MOUSEEVENTF_VIRTUALDESK = 0x4000

antecanis8 committed:
fix: 移除ZContext类中win_title参数

antecanis8 committed:
fix: 跳过插队或排队, 直接进入游戏

ShadowLemoon commented on src/zzz_od/operation/enter_game/enter_game.py:
```diff
@@ -35,6 +36,9 @@ def __init__(self, ctx: ZContext, switch: bool = False):
             self.force_login = True
 
         self.already_login: bool = False  # 是否已经登录了
+        if self.ctx.game_account_config.is_cloud_game:
```
有没有场景会复用这个操作实例？如果有的话放在handle_init内部比较好

ShadowLemoon commented on src/zzz_od/gui/view/setting/setting_game_interface.py:
```diff
@@ -66,12 +66,25 @@ def get_content_widget(self) -> QWidget:
         return content_widget
 
+    def _get_cloud_game_group(self) -> QWidget:
+        cloud_game_group = SettingCardGroup(gt('云游戏设置'))
```
丢基础下面吧 可以只叫云游戏去掉设置两字

ShadowLemoon commented on src/zzz_od/context/zzz_context.py:
```diff
@@ -91,16 +91,25 @@ def reload_instance_config(self) -> None:
     def _get_win_title(self) -> str:
         """获取当前配置对应的窗口标题"""
-        if self.game_account_config.use_custom_win_title:
+        if self.game_account_config.use_custom_win_title \
```
别用\转义换行

ShadowLemoon commented on src/one_dragon/base/config/game_account_config.py:
```diff
@@ -63,13 +75,54 @@ def custom_win_title(self) -> str:
+    def is_cloud_game(self) -> bool:
+        return self.client_type == ClientTypeEnum.CLOUD.value.value
+
+    @property
```
没挪走

ShadowLemoon commented on src/one_dragon/base/controller/pc_controller_base.py:
```diff
@@ -563,16 +565,85 @@ def get_mouse_sensitivity():
 
-def drag_mouse(start: Point, end: Point, duration: float = 0.5):
+def drag_mouse(
+    start: Point,
```
这个函数main分支有更改，需要重新适配

antecanis8 committed:
Merge remote-tracking branch 'upstream/main' into cloud2

# Conflicts:
#	src/one_dragon/base/config/game_account_config.py
#	src/one_dragon_qt/view/setting/setting_instance_interface.py
#	src/zzz_od/operation/enter_game/enter_game.py
#	src/zzz_od/operation/enter_game/open_and_enter_game.py

antecanis8 committed:
Merge remote-tracking branch 'upstream/main' into cloud2

# Conflicts:
#	src/one_dragon/base/controller/pc_controller_base.py
#	src/zzz_od/controller/zzz_pc_controller.py

antecanis8 committed:
feat(controller): 添加设置窗口句柄功能以支持业务层确认

antecanis8 committed:
feat(cloud_game): 添加云游戏窗口选择器以优化截图流程

新增 CloudGameWindowSelector 类，支持从多个同名窗口中选择有效窗口。
改进截图逻辑，确保在云游戏排队时选择正确的 HWND，
并在窗口探测失败时实现超时处理，提升用户体验。
